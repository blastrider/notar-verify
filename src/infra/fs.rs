use anyhow::{bail, Context, Result};
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Clone, Copy, Debug)]
pub struct Limits {
    pub max_bytes: u64,
}

impl Limits {
    pub fn from_mib(mib: u64) -> Self {
        Self { max_bytes: mib * 1024 * 1024 }
    }
}

fn normalize(path: &str) -> Result<PathBuf> {
    let pb = PathBuf::from(path);
    let abs = pb.canonicalize().with_context(|| format!("Chemin invalide: {path}"))?;
    let s = abs.to_string_lossy();
    if s.contains("..") {
        bail!("Traversal détecté");
    }
    Ok(abs)
}

pub fn read_file_bounded(path: &str, limits: &Limits) -> Result<Vec<u8>> {
    let p = normalize(path)?;
    let md = fs::metadata(&p).with_context(|| format!("Stat échouée: {}", p.display()))?;
    if md.len() > limits.max_bytes {
        bail!("Fichier trop volumineux ({} bytes > limite)", md.len());
    }
    let data = fs::read(&p).with_context(|| format!("Lecture échouée: {}", p.display()))?;
    Ok(data)
}

pub fn read_all_pems(paths: &[String], limits: &Limits) -> Result<Vec<String>> {
    let mut res = Vec::new();
    for p in paths {
        let d = read_file_bounded(p, limits)?;
        let s = String::from_utf8(d).context("PEM non UTF-8")?;
        // Check minimal PEM header presence
        if !s.contains("-----BEGIN") {
            bail!("Fichier --trust sans bloc PEM BEGIN/END");
        }
        res.push(s);
    }
    Ok(res)
}
