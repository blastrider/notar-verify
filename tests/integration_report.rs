use assert_cmd::Command;
use std::fs;
use tempfile::tempdir;

#[test]
fn cli_shows_warning_without_backend() {
    let dir = tempdir().unwrap();
    let data = dir.path().join("data.bin");
    let sig = dir.path().join("detached.sig.p7s");
    fs::write(&data, b"hello world").unwrap();
    fs::write(&sig, b"not-a-real-p7s").unwrap();
    let out = dir.path().join("report.json");

    let mut cmd = Command::cargo_bin("notar-verify").unwrap();
    let assert = cmd
        .arg("--sig")
        .arg(sig)
        .arg("--data")
        .arg(data)
        .arg("--out")
        .arg(&out)
        .assert();

    // Ne pas exiger un exit code 0 : selon la build/features, le binaire peut retourner
    // 1/2 (Invalid/Warning) tout en ayant écrit le rapport JSON. Vérifions seulement
    // que le fichier a bien été créé avant de le lire.
    assert!(
        out.exists(),
        "Le rapport JSON attendu n'a pas été créé : {:?}",
        out
    );

    let json = fs::read_to_string(out).unwrap();
    insta::assert_snapshot!("report_warning", json);
}
