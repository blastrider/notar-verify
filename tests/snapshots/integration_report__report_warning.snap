---
source: tests/integration_report.rs
expression: json
---
{
  "input_kind": "CMS",
  "algorithms": [],
  "signer_dn": null,
  "certificate_chain": [],
  "signing_time": null,
  "timestamp_rfc3161": null,
  "revocation": {
    "status": "WARNING",
    "detail": "Révocation offline non évaluée (MVP). Fournir CRL/OCSP et activer backend ultérieurement."
  },
  "integrity": {
    "status": "WARNING",
    "detail": ""
  },
  "signature": {
    "status": "WARNING",
    "detail": "Pile native CMS non activée : utilisez --features openssl-backend pour une vérification cryptographique complète."
  },
  "chain": {
    "status": "WARNING",
    "detail": "Validation de chaîne limitée sans backend X.509 avancé."
  },
  "ltv": {
    "status": "WARNING",
    "detail": ""
  },
  "verdict": "WARNING",
  "document_sha256": "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9"
}
